fix836_getLoadedLibraries2 (second effort at PR).

clanmills opened the pull request.

clanmills committed:
fix836_getLoadedLibraries2 (second effort at PR).

piponazo commented on src/futils.cpp:
```diff
@@ -502,4 +493,71 @@ namespace Exiv2 {
+
+    #elif defined(__unix__)
+        // read file /proc/self/maps which has a list of files in memory
+        std::ifstream maps("/proc/self/maps",std::ifstream::in);
```
You will need to include the header `<fstream>`, right now the linux jobs are failing with this error:

```
/home/travis/build/Exiv2/exiv2/src/futils.cpp: In function ‘std::vector<std::__cxx11::basic_string<char> > Exiv2::getLoadedLibraries()’:
/home/travis/build/Exiv2/exiv2/src/futils.cpp:551:27: error: variable ‘std::ifstream maps’ has initializer but incomplete type
         std::ifstream maps("/proc/self/maps",std::ifstream::in);
```

piponazo commented on src/version.cpp:
```diff
@@ -412,7 +308,7 @@ void Exiv2::dumpLibraryInfo(std::ostream& os,const exv_grep_keys_t& keys)
     if ( libs.begin() != libs.end() ) {
         output(os,keys,"executable" ,*libs.begin());
-        for ( Exiv2::StringVector_i lib = libs.begin()+1 ; lib != libs.end() ; ++lib )
+        for ( std::vector<std::string>::iterator lib = libs.begin()+1 ; lib != libs.end() ; ++lib )
```
Here we could use `auto` for not typing the long iterator type:

```cpp
for (const auto& lib = libs.begin()+1 ; lib != libs.end() ; ++lib )
```

clanmills replied:
That doesn't compile on the Mac with Clang.

clanmills committed:
Moved `#include <fstream>` from version.cpp to futils.cpp

piponazo commented:
Windows builds are failing with this error:

```
..\src\futils.cpp(519): error C3861: 'lengthof': identifier not found
..\src\futils.cpp(522): error C3861: 'lengthof': identifier not found
```

And the Fedora_MinGW one with this one:

```
src/futils.cpp:518:61: error: 'lengthof' was not declared in this scope
         if ( EnumProcessModules(GetCurrentProcess(),handles,lengthof(handles),&cbNeeded)) {
```

D4N commented on src/futils.cpp:
```diff
@@ -502,4 +494,71 @@ namespace Exiv2 {
+        // enumerate loaded libraries and determine path to executable
+        HMODULE handles[200];
+        DWORD   cbNeeded;
+        if ( EnumProcessModules(GetCurrentProcess(),handles,lengthof(handles),&cbNeeded)) {
```
```suggestion
        if ( EnumProcessModules(GetCurrentProcess(),handles,EXV_COUNTOF(handles),&cbNeeded)) {
```

D4N commented on src/futils.cpp:
```diff
@@ -502,4 +494,71 @@ namespace Exiv2 {
+        if ( EnumProcessModules(GetCurrentProcess(),handles,lengthof(handles),&cbNeeded)) {
+            char szFilename[_MAX_PATH];
+            for ( DWORD h = 0 ; h < cbNeeded/sizeof(handles[0]) ; h++ ) {
+                GetModuleFileNameA(handles[h],szFilename,lengthof(szFilename)) ;
```
```suggestion
                GetModuleFileNameA(handles[h],szFilename,EXV_COUNTOF(szFilename)) ;
```

clanmills replied:
You're right Dan.  That's what I should have done.  I added macro lengthof() to src/futils.cpp

D4N reviewed (commented):
Replacing `lengthof` with `EXV_COUNTOF` should do the trick.

clanmills committed:
Add lengthof() macro to src/futils.cpp

piponazo committed:
Better platform protection for uid_t types

We were obtaining this error in CI:
```
exiv2/src/version.cpp:338:5: error: 'uid_t' was not declared in this scope
     uid_t uid  = getuid()  ; output(os,keys,"uid" ,  uid  );
     ^~~~~
exiv2/src/version.cpp:338:5: note: suggested alternative: 'pid_t'
     uid_t uid  = getuid()  ; output(os,keys,"uid" ,  uid  );
     ^~~~~
```
